File: relationship-manager/conversation-service/src/main/java/com/bank/rm/conversation/service/ConversationService.java
package com.bank.rm.conversation.service;

import com.bank.rm.common.dto.*;
import com.bank.rm.common.event.*;
import com.bank.rm.common.exception.ResourceNotFoundException;
import com.bank.rm.conversation.ai.ConversationAIAgent;
import com.bank.rm.conversation.ai.EntityExtractor;
import com.bank.rm.conversation.domain.Conversation;
import com.bank.rm.conversation.domain.Message;
import com.bank.rm.conversation.dto.ConversationDtos.*;
import com.bank.rm.conversation.engine.ConversationStateMachine;
import com.bank.rm.conversation.repository.ConversationRepository;
import com.bank.rm.conversation.repository.MessageRepository;
import com.fasterxml.jackson.core.type.TypeReference;
import com.fasterxml.jackson.databind.ObjectMapper;
import org.slf4j.Logger;
import org.slf4j.LoggerFactory;
import org.springframework.kafka.core.KafkaTemplate;
import org.springframework.stereotype.Service;
import org.springframework.transaction.annotation.Transactional;
import java.time.Instant;
import java.util.*;
import java.util.stream.Collectors;

@Service
public class ConversationService {
    private static final Logger log = LoggerFactory.getLogger(ConversationService.class);
    private final ConversationRepository conversationRepository;
    private final MessageRepository messageRepository;
    private final ConversationAIAgent aiAgent;
    private final EntityExtractor entityExtractor;
    private final ConversationStateMachine stateMachine;
    private final KafkaTemplate<String, String> kafkaTemplate;
    private final ObjectMapper objectMapper;

    public ConversationService(ConversationRepository conversationRepository,
                                MessageRepository messageRepository,
                                ConversationAIAgent aiAgent,
                                EntityExtractor entityExtractor,
                                ConversationStateMachine stateMachine,
                                KafkaTemplate<String, String> kafkaTemplate,
                                ObjectMapper objectMapper) {
        this.conversationRepository = conversationRepository;
        this.messageRepository = messageRepository;
        this.aiAgent = aiAgent;
        this.entityExtractor = entityExtractor;
        this.stateMachine = stateMachine;
        this.kafkaTemplate = kafkaTemplate;
        this.objectMapper = objectMapper;
    }

    @Transactional
    public ConversationResponse startConversation(StartConversationRequest request) {
        UUID conversationId = UUID.randomUUID();
        CommunicationChannel channel = request.channel() != null ? request.channel() : CommunicationChannel.WEB;
        Conversation conversation = new Conversation(
                conversationId, request.customerId(), request.anonymousSessionId(), channel);
        conversationRepository.save(conversation);

        String greeting = aiAgent.generateResponse("", ConversationPhase.GREETING, Map.of(), "");
        Message greetingMsg = new Message(UUID.randomUUID(), conversationId, SenderType.AI, null,
                ContentType.TEXT, greeting, ConversationPhase.GREETING);
        messageRepository.save(greetingMsg);

        publishEvent(KafkaTopics.CONVERSATION_EVENTS,
                new ConversationStartedEvent(conversationId, request.customerId(),
                        request.anonymousSessionId(), channel));

        return new ConversationResponse(conversationId, conversation.getCurrentPhase(),
                conversation.getStatus(), channel, conversation.getCreatedAt(), Instant.now());
    }

    @Transactional
    @SuppressWarnings("unchecked")
    public SendMessageResponse sendMessage(UUID conversationId, SendMessageRequest request) {
        Conversation conversation = conversationRepository.findById(conversationId)
                .orElseThrow(() -> new ResourceNotFoundException("Conversation not found: " + conversationId));

        Message customerMsg = new Message(UUID.randomUUID(), conversationId, SenderType.CUSTOMER,
                conversation.getCustomerId(), ContentType.TEXT, request.content(), conversation.getCurrentPhase());
        messageRepository.save(customerMsg);

        Map<String, Object> existingData = parseJson(conversation.getExtractedData());
        Map<String, Object> updatedData = entityExtractor.extract(
                request.content(), conversation.getCurrentPhase(), existingData);
        conversation.setExtractedData(toJson(updatedData));

        if (stateMachine.canTransition(conversation.getCurrentPhase(), updatedData)) {
            ConversationPhase nextPhase = stateMachine.getNextPhase(conversation.getCurrentPhase());
            conversation.setCurrentPhase(nextPhase);
            if (nextPhase == ConversationPhase.COMPLETED) {
                conversation.setStatus(ConversationStatus.COMPLETED);
                String summary = aiAgent.generateSummary(getHistory(conversationId), updatedData);
                conversation.setSummary(summary);
                publishEvent(KafkaTopics.CONVERSATION_EVENTS,
                        new ConversationCompletedEvent(conversationId, conversation.getCustomerId(),
                                summary, List.of(ConversationPhase.values())));
            }
        }

        conversation.setUpdatedAt(Instant.now());
        conversationRepository.save(conversation);

        String history = getHistory(conversationId);
        String aiResponse = aiAgent.generateResponse(request.content(), conversation.getCurrentPhase(),
                updatedData, history);
        Message aiMsg = new Message(UUID.randomUUID(), conversationId, SenderType.AI, null,
                ContentType.TEXT, aiResponse, conversation.getCurrentPhase());
        messageRepository.save(aiMsg);

        return new SendMessageResponse(toMessageResponse(customerMsg), toMessageResponse(aiMsg));
    }

    public ConversationDetailResponse getConversation(UUID conversationId) {
        Conversation conversation = conversationRepository.findById(conversationId)
                .orElseThrow(() -> new ResourceNotFoundException("Conversation not found: " + conversationId));
        List<Message> messages = messageRepository.findByConversationIdOrderByCreatedAtAsc(conversationId);
        List<MessageResponse> messageResponses = messages.stream()
                .map(this::toMessageResponse).collect(Collectors.toList());
        Map<String, Object> extractedData = parseJson(conversation.getExtractedData());
        return new ConversationDetailResponse(conversationId, conversation.getCurrentPhase(),
                conversation.getStatus(), messageResponses, extractedData);
    }

    private String getHistory(UUID conversationId) {
        List<Message> messages = messageRepository.findByConversationIdOrderByCreatedAtAsc(conversationId);
        StringBuilder sb = new StringBuilder();
        for (Message m : messages) {
            sb.append(m.getSenderType() == SenderType.CUSTOMER ? "Customer: " : "RM: ")
              .append(m.getContent()).append("\n");
        }
        return sb.toString();
    }

    @SuppressWarnings("unchecked")
    private MessageResponse toMessageResponse(Message m) {
        Map<String, Object> entities = parseJson(m.getEntitiesExtracted());
        return new MessageResponse(m.getId(), m.getConversationId(), m.getSenderType(),
                m.getContent(), m.getContentType(), m.getPhase(), entities, m.getCreatedAt());
    }

    @SuppressWarnings("unchecked")
    private Map<String, Object> parseJson(String json) {
        try {
            if (json == null || json.isEmpty()) return new HashMap<>();
            return objectMapper.readValue(json, new TypeReference<Map<String, Object>>() {});
        } catch (Exception e) {
            return new HashMap<>();
        }
    }

    private String toJson(Object obj) {
        try {
            return objectMapper.writeValueAsString(obj);
        } catch (Exception e) {
            return "{}";
        }
    }

    private void publishEvent(String topic, Object event) {
        try {
            kafkaTemplate.send(topic, toJson(event));
        } catch (Exception e) {
            log.error("Failed to publish event to {}: {}", topic, e.getMessage());
        }
    }
}

File: relationship-manager/conversation-service/src/main/kotlin/com/bank/rm/conversation/service/ConversationService.kt
package com.bank.rm.conversation.service

import com.bank.rm.common.dto.*
import com.bank.rm.common.event.*
import com.bank.rm.common.exception.ResourceNotFoundException
import com.bank.rm.common.exception.ValidationException
import com.bank.rm.conversation.ai.ChatMessage
import com.bank.rm.conversation.ai.ConversationAIAgent
import com.bank.rm.conversation.domain.*
import com.bank.rm.conversation.dto.*
import com.bank.rm.conversation.engine.ConversationStateMachine
import com.bank.rm.conversation.repository.*
import com.fasterxml.jackson.databind.ObjectMapper
import com.fasterxml.jackson.module.kotlin.readValue
import org.slf4j.LoggerFactory
import org.springframework.kafka.core.KafkaTemplate
import org.springframework.stereotype.Service
import org.springframework.transaction.annotation.Transactional
import java.time.Instant
import java.util.UUID

@Service
class ConversationService(
    private val conversationRepository: ConversationRepository,
    private val messageRepository: MessageRepository,
    private val aiAgent: ConversationAIAgent,
    private val stateMachine: ConversationStateMachine,
    private val kafkaTemplate: KafkaTemplate<String, DomainEvent>,
    private val objectMapper: ObjectMapper
) {
    private val log = LoggerFactory.getLogger(javaClass)

    @Transactional
    fun startConversation(request: StartConversationRequest): ConversationResponse {
        val conversation = Conversation(
            customerId = request.customerId,
            anonymousSessionId = request.anonymousSessionId ?: UUID.randomUUID(),
            channel = request.channel,
            currentPhase = ConversationPhase.GREETING,
            status = ConversationStatus.ACTIVE
        )
        conversationRepository.save(conversation)

        // Generate greeting using AI
        val greeting = try {
            aiAgent.generateResponse(
                ConversationPhase.GREETING,
                emptyList(),
                emptyMap(),
                "[NEW_CONVERSATION]"
            )
        } catch (e: Exception) {
            log.warn("AI greeting failed, using fallback", e)
            null
        }

        val greetingText = greeting?.message
            ?: "Welcome! I'm your digital Relationship Manager. I'm here to help you explore financial products tailored to your goals. How can I assist you today?"

        val greetingMsg = Message(
            conversationId = conversation.id,
            senderType = SenderType.AI,
            contentType = ContentType.TEXT,
            content = greetingText,
            phase = ConversationPhase.GREETING
        )
        messageRepository.save(greetingMsg)

        kafkaTemplate.send(
            KafkaTopics.CONVERSATION_EVENTS,
            conversation.id.toString(),
            ConversationStartedEvent(
                conversationId = conversation.id,
                customerId = request.customerId,
                anonymousSessionId = conversation.anonymousSessionId,
                channel = request.channel
            )
        )

        return conversation.toResponse()
    }

    @Transactional
    fun sendMessage(conversationId: UUID, request: SendMessageRequest, senderId: UUID?): Pair<MessageResponse, MessageResponse> {
        val conversation = findConversation(conversationId)
        if (conversation.status != ConversationStatus.ACTIVE) {
            throw ValidationException("Conversation is not active: ${conversation.status}")
        }

        // Save customer message
        val customerMsg = Message(
            conversationId = conversationId,
            senderType = SenderType.CUSTOMER,
            senderId = senderId,
            contentType = request.contentType,
            content = request.content,
            phase = conversation.currentPhase
        )
        messageRepository.save(customerMsg)

        // Check for handoff intent
        if (aiAgent.detectHandoffIntent(request.content)) {
            return handleHandoff(conversation, customerMsg)
        }

        // Get conversation history for context
        val history = messageRepository.findByConversationIdOrderByCreatedAtAsc(conversationId)
            .map { ChatMessage(if (it.senderType == SenderType.CUSTOMER) "Customer" else "RM", it.content) }

        val extractedData: Map<String, Any> = try {
            objectMapper.readValue(conversation.extractedData)
        } catch (e: Exception) { emptyMap() }

        // Generate AI response
        val aiResponse = aiAgent.generateResponse(
            conversation.currentPhase,
            history,
            extractedData,
            request.content
        )

        // Merge extracted entities
        val updatedData = extractedData + aiResponse.extractedEntities
        conversation.extractedData = objectMapper.writeValueAsString(updatedData)

        // Phase transition
        if (aiResponse.shouldTransitionPhase && aiResponse.suggestedNextPhase != null) {
            if (stateMachine.canTransition(conversation.currentPhase, aiResponse.suggestedNextPhase)) {
                conversation.currentPhase = aiResponse.suggestedNextPhase
                conversation.status = stateMachine.determineStatus(aiResponse.suggestedNextPhase)
            }
        }

        conversation.lastMessageAt = Instant.now()
        conversation.updatedAt = Instant.now()
        conversationRepository.save(conversation)

        // Save AI response message
        val aiMsg = Message(
            conversationId = conversationId,
            senderType = SenderType.AI,
            contentType = ContentType.TEXT,
            content = aiResponse.message,
            phase = conversation.currentPhase,
            entitiesExtracted = objectMapper.writeValueAsString(aiResponse.extractedEntities)
        )
        messageRepository.save(aiMsg)

        // If conversation completed, emit event
        if (conversation.status == ConversationStatus.COMPLETED) {
            emitCompletionEvent(conversation, updatedData)
        }

        return Pair(customerMsg.toResponse(), aiMsg.toResponse())
    }

    fun getConversationHistory(conversationId: UUID): ConversationHistoryResponse {
        val conversation = findConversation(conversationId)
        val messages = messageRepository.findByConversationIdOrderByCreatedAtAsc(conversationId)
        val extractedData: Map<String, Any> = try {
            objectMapper.readValue(conversation.extractedData)
        } catch (e: Exception) { emptyMap() }

        return ConversationHistoryResponse(
            conversationId = conversation.id,
            currentPhase = conversation.currentPhase,
            status = conversation.status,
            messages = messages.map { it.toResponse() },
            extractedData = extractedData
        )
    }

    fun getCustomerConversations(customerId: UUID): List<ConversationResponse> =
        conversationRepository.findByCustomerIdOrderByLastMessageAtDesc(customerId)
            .map { it.toResponse() }

    @Transactional
    fun summarizeConversation(conversationId: UUID): ConversationSummaryResponse {
        val conversation = findConversation(conversationId)
        val messages = messageRepository.findByConversationIdOrderByCreatedAtAsc(conversationId)
        val extractedData: Map<String, Any> = try {
            objectMapper.readValue(conversation.extractedData)
        } catch (e: Exception) { emptyMap() }

        val history = messages.map { ChatMessage(if (it.senderType == SenderType.CUSTOMER) "Customer" else "RM", it.content) }
        val summary = aiAgent.generateSummary(history, extractedData)

        conversation.summary = summary
        conversationRepository.save(conversation)

        val phasesCompleted = ConversationPhase.entries.filter { it.ordinal <= conversation.currentPhase.ordinal }

        return ConversationSummaryResponse(
            conversationId = conversation.id,
            summary = summary,
            phasesCompleted = phasesCompleted,
            extractedData = extractedData
        )
    }

    private fun handleHandoff(conversation: Conversation, customerMsg: Message): Pair<MessageResponse, MessageResponse> {
        conversation.status = ConversationStatus.HANDED_OFF
        conversationRepository.save(conversation)

        val aiMsg = Message(
            conversationId = conversation.id,
            senderType = SenderType.AI,
            contentType = ContentType.TEXT,
            content = "I understand you'd like to speak with a human relationship manager. Let me connect you right away. A team member will be in touch shortly. Thank you for your patience!",
            phase = conversation.currentPhase
        )
        messageRepository.save(aiMsg)

        return Pair(customerMsg.toResponse(), aiMsg.toResponse())
    }

    private fun emitCompletionEvent(conversation: Conversation, extractedData: Map<String, Any>) {
        kafkaTemplate.send(
            KafkaTopics.CONVERSATION_EVENTS,
            conversation.id.toString(),
            ConversationCompletedEvent(
                conversationId = conversation.id,
                customerId = conversation.customerId ?: UUID.randomUUID(),
                summary = conversation.summary,
                phasesCompleted = ConversationPhase.entries.toList()
            )
        )
    }

    private fun findConversation(conversationId: UUID): Conversation =
        conversationRepository.findById(conversationId).orElseThrow {
            ResourceNotFoundException("Conversation not found: $conversationId")
        }

    private fun Conversation.toResponse() = ConversationResponse(
        id, currentPhase, status, channel, createdAt.toString(), lastMessageAt.toString()
    )

    private fun Message.toResponse(): MessageResponse {
        val entities: Map<String, Any>? = entitiesExtracted?.let {
            try { objectMapper.readValue(it) } catch (e: Exception) { null }
        }
        return MessageResponse(id, conversationId, senderType, content, contentType, phase, entities, createdAt.toString())
    }
}
